File: utils/security.py
import discord
from discord.ext import commands
from config import ALLOWED_USER_IDS, BLOCKED_COMMANDS


def is_allowed_user():
    """Check decorator: only whitelisted users can use commands."""

    async def predicate(ctx: commands.Context) -> bool:
        if not ALLOWED_USER_IDS:
            return True
        if ctx.author.id not in ALLOWED_USER_IDS:
            await ctx.reply("You are not authorized to use this bot.")
            return False
        return True

    return commands.check(predicate)


def is_command_blocked(command: str) -> bool:
    """Check if a CMD command is in the blocklist."""
    cmd_lower = command.strip().lower()
    for blocked in BLOCKED_COMMANDS:
        if cmd_lower.startswith(blocked):
            return True
    return False

File: src/utils/security.ts
import { ALLOWED_USER_IDS, BLOCKED_COMMANDS } from "../config.js";

/** Check if a Discord user is whitelisted. Empty whitelist = deny all. */
export function isAllowedUser(userId: string): boolean {
  if (ALLOWED_USER_IDS.size === 0) {
    console.warn("[security] ALLOWED_USER_IDS is empty — all access denied. Set user IDs in .env.");
    return false;
  }
  return ALLOWED_USER_IDS.has(userId);
}

/**
 * Dangerous executable patterns that should be blocked even when
 * they appear in the middle of a command (e.g. piped or chained).
 */
const BLOCKED_PATTERNS: RegExp[] = [
  /\bpowershell(?:\.exe)?\b/i,
  /\bpwsh(?:\.exe)?\b/i,
  /\bcmd(?:\.exe)?\s*\/c\b/i,
  /\bwscript(?:\.exe)?\b/i,
  /\bcscript(?:\.exe)?\b/i,
  /\bmshta(?:\.exe)?\b/i,
  /\bcertutil(?:\.exe)?\b/i,
  /\bbitsadmin(?:\.exe)?\b/i,
  /\bwmic(?:\.exe)?\b/i,
  /\bregsvr32(?:\.exe)?\b/i,
  /\brundll32(?:\.exe)?\b/i,
];

/** Check if a CMD command is in the blocklist. */
export function isCommandBlocked(command: string): boolean {
  const lower = command.trim().toLowerCase();

  // Prefix match against BLOCKED_COMMANDS set
  for (const blocked of BLOCKED_COMMANDS) {
    if (lower.startsWith(blocked)) return true;
  }

  // Regex pattern match for dangerous executables anywhere in command
  for (const pattern of BLOCKED_PATTERNS) {
    if (pattern.test(command)) return true;
  }

  return false;
}
